scipy/stats/_morestats.py

```diff
@@ -3293,78 +3293,69 @@ def func(x):
-    S = np.cumsum(t)
-    # define a copy of `S` with a prepending zero for later use to avoid
-    # the need for indexing.
-    S_i_m1 = np.concatenate(([0], S[:-1]))
-
-    # Psi, as defined by the 6th unnumbered equation on page 313 (Mielke).
-    # Note that in the paper there is an error where the denominator `2` is
-    # squared when it should be the entire equation.
-    def psi(indicator):
-        return (indicator - (N + 1)/2)**2
-
-    # define summation range for use in calculation of phi, as seen in sum
-    # in the unnumbered equation on the bottom of page 312 (Mielke).
-    s_lower = S[js - 1] + 1
-    s_upper = S[js] + 1
-    phi_J = [np.arange(s_lower[idx], s_upper[idx]) for idx in range(k)]
-
-    # for every range in the above array, determine the sum of psi(I) for
-    # every element in the range. Divide all the sums by `t`. Following the
-    # last unnumbered equation on page 312.
-    phis = [np.sum(psi(I_j)) for I_j in phi_J] / t[js]
-
-    # `T` is equal to a[j] * phi[j], per the first unnumbered equation on
-    # page 312. `phis` is already in the order based on `js`, so we index
-    # into `a` with `js` as well.
-    T = sum(phis * a[js])
-
-    # The approximate statistic
+def _mood_statistic_with_ties(x, y, t, n, m, N):
+    # First equation of "Mood's Squared Rank Test", Mielke pg 313
     E_0_T = n * (N * N - 1) / 12
 
-    varM = (m * n * (N + 1.0) * (N ** 2 - 4) / 180 -
-            m * n / (180 * N * (N - 1)) * np.sum(
-                t * (t**2 - 1) * (t**2 - 4 + (15 * (N - S - S_i_m1) ** 2))
-            ))
-
-    return ((T - E_0_T) / np.sqrt(varM),)
+    # m, n, N, t, and S are defined in the second paragraph of Mielke pg 312
+    # The only difference is that our `t` has zeros interspersed with the relevant
+    # numbers to keep the array rectangular, but these terms nothing to the sum.
+    np = array_namespace(x, y, t)  # needed to use `cumulative_sum`
+    S = np.cumulative_sum(t, include_initial=True, axis=-1)
+    S_i, S_i_m1 = S[..., 1:], S[..., :-1]
+    # Second equation of "Mood's Squared Rank Test", Mielke pg 313
+    varM = (m * n * (N + 1.0) * (N**2 - 4) / 180
+            - m * n / (180 * N * (N - 1))
+            * np.sum(t * (t ** 2 - 1) * (t ** 2 - 4 + (15 * (N - S_i - S_i_m1) ** 2)),
+                     axis=-1))
+
+    # There is a formula for phi in terms of t, S, and Psi(I) = [I - (N+1)/2]^2 is
+    # defined (with a mistake in the location of the ^2) at the beginning of
+    # "Mood's Squared Rank Test", Mielke pg 313. To vectorize this calculation,
+    # let c = (N + 1) / 2, so Psi(I) = I^2 - 2*c*I + c^2. We sum each of these three
+    # parts of Psi separately using formulas for sums from a to b (inclusive) of
+    # terms I^2, I, and 1 where I takes on successive integers.
+    def sum_I2(a, b):
+        return (b * (b + 1) * (2*b + 1) - a * (a - 1) * (2*a - 1)) / 6
+
+    def sum_I(a, b):
+        return (b * (b + 1) - a * (a - 1)) / 2
```

Comment:
The sum of consecutive integers from `a` to `b`, inclusive. The inclusivity is why there's a minus sign instead of a plus in the second term.
```python3
from sympy import *
a, b = symbols('a, b')
x = b * (b + 1) / 2 - a * (a + 1) / 2 + a
y = (b * (b + 1) - a * (a - 1)) / 2
simplify(x - y)
```